Philips.CodeAnalysis.MsTestAnalyzers/TestHasTimeoutCodeFixProvider.cs

```diff
@@ -0,0 +1,105 @@
+﻿using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+using Philips.CodeAnalysis.Common;
+
+namespace Philips.CodeAnalysis.MsTestAnalyzers
+{
+	[ExportCodeFixProvider(LanguageNames.CSharp, Name = nameof(TestHasTimeoutCodeFixProvider)), Shared]
+	public class TestHasTimeoutCodeFixProvider : CodeFixProvider
+	{
+		private const string Title = "Add Test Timeout";
+
+		public sealed override ImmutableArray<string> FixableDiagnosticIds
+		{
+			get { return ImmutableArray.Create(Helper.ToDiagnosticId(DiagnosticIds.TestHasTimeoutAttribute)); }
+		}
+
+		public sealed override FixAllProvider GetFixAllProvider()
+		{
+			// See https://github.com/dotnet/roslyn/blob/master/docs/analyzers/FixAllProvider.md for more information on Fix All Providers
+			return WellKnownFixAllProviders.BatchFixer;
+		}
+
+		public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+		{
+			SyntaxNode root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+			Diagnostic diagnostic = context.Diagnostics.First();
+			TextSpan diagnosticSpan = diagnostic.Location.SourceSpan;
+
+			// Find the method declaration identified by the diagnostic.
+			MethodDeclarationSyntax attributeList = root.FindToken(diagnosticSpan.Start).Parent.AncestorsAndSelf().OfType<MethodDeclarationSyntax>().First();
+
+			// Register a code action that will invoke the fix.
+			context.RegisterCodeFix(
+				CodeAction.Create(
+					title: Title,
+					createChangedDocument: c => AddTestTimeout(context.Document, attributeList, c),
+					equivalenceKey: Title),
+				diagnostic);
+		}
+
+		private async Task<Document> AddTestTimeout(Document document, MethodDeclarationSyntax method, CancellationToken cancellationToken)
+		{
+			SyntaxList<AttributeListSyntax> attributeLists = method.AttributeLists;
+			string category = string.Empty;
+			string timeout = string.Empty;
+			foreach (AttributeListSyntax attributes in attributeLists)
+			{
+				foreach (AttributeSyntax attributesyntax in attributes.Attributes)
+				{
+					if (attributesyntax.Name.ToString().Contains(@"Timeout"))
+					{
+						return document;
+					}
+					else if (attributesyntax.Name.ToString().Contains(@"TestCategory"))
+					{
+						category = attributesyntax.ArgumentList.Arguments.First().ToString();
+					}
+				}
+			}
+
+			switch (category)
+			{
```

Comment:
These are arguably weird values for open source.

Reply:
Agreed.  Let me think about it.

Reply:
Changed behavior.  Categories and acceptable test timeouts are listed in the editorconfig.  If you have the analyzer enabled, but have a category not in the editorconfig it sets the timeout to 1000.